### PR TITLE
CDD-2174: Time series and Headline Audit endpoints

### PR DESCRIPTION
**Pull request by phill-stanley**
# Description

This PR includes endpoints to return all records for a given slice of data including embargoed data and stale data for auditing and debugging, this includes endpoints for the following: 

- API time series
- Core time series
- Core headline

Fixes #CDD-2174

---

## Type of change

Please select the options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Tech debt item (this is focused solely on addressing any relevant technical debt)

---

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests at the right levels to prove my change is effective
- [ ] I have added screenshots or screen grabs where appropriate
- [x] I have added docstrings in the correct style [(google)](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings)

### CODE REVIEW

File: metrics/api/views/audit/serializers/audit_core_headline.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    period_end = serializers.SerializerMethodField()
+    refresh_date = serializers.SerializerMethodField()
+    embargo = serializers.SerializerMethodField()
+
+    class Meta:
+        model = CoreHeadline
+        fields = "__all__"
+
+    @classmethod
+    def get_topic(cls, obj) -> str:
```
Comment (A-Ashiq) [inline metrics/api/views/audit/serializers/audit_core_headline.py:36]: Reckon it's worth type hinting all of these `obj` with `obj: CoreHeadline`?

File: metrics/api/views/audit/serializers/audit_core_headline.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    def get_age(cls, obj) -> str:
+        return obj.age.name
+
+    @classmethod
+    def get_stratum(cls, obj) -> str:
+        return obj.stratum.name
+
+    @classmethod
+    def get_period_start(cls, obj) -> str:
+        return datetime.strftime(obj.period_start, "%Y-%m-%d %H:%M:%S")
```
Comment (A-Ashiq) [inline metrics/api/views/audit/serializers/audit_core_headline.py:73]: Probably worth making a constant at the top of this file:
```
EXPECTED_TIMESTAMP_FORMAT = "%Y-%m-%d %H:%M:%S"
```
And then just referencing that in these methods

File: metrics/api/views/audit/audit_core_headline.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    """
+
+    permission_classes = []
+    queryset = CoreHeadline.objects.all()
+    serializer_class = AuditCoreHeadlineSerializer
+    pagination_class = AuditEndpointPagination
+    filter_backends = [DjangoFilterBackend]
+
+    def get_queryset(self) -> CoreTimeSeriesQuerySet:
```
Comment (A-Ashiq) [inline metrics/api/views/audit/audit_core_headline.py:75]: This is headline not time series

File: metrics/api/views/audit/audit_core_timeseries.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    - stratum `default`
+
+    This would be a **valid** request because the `metric` of `COVID-19_cases_casesByDay` relates to `Timeseries` data.
+
+    Note the key difference here being the 2nd part of the metric naming:
+
+    *<topic>* *<metric group>* *<description>*
+
+    A valid metric name for this endpoint should **NOT** include `Headline` as the metric group part of the name.
```
Comment (A-Ashiq) [inline metrics/api/views/audit/audit_core_timeseries.py:66]:
```suggestion
    A valid metric name for this endpoint should **NOT** include `headline` as the metric group part of the name.
```
I'd be specific about the case-sensitivity here

File: metrics/api/views/audit/audit_core_headline.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    - stratum `default`
+
+    This would be **valid** because the `metric` of `COVID-19_headline_ONSdeaths_7DayChange` relates to headline data.
+
+    Note the key difference here being the 2nd part of the metric naming:
+
+    *<topic>* *<metric group>* *<description>*
+
+    A valid metric name for this endpoint should include `Headline` as the metric group part of the name.
```
Comment (A-Ashiq) [inline metrics/api/views/audit/audit_core_headline.py:65]:
```suggestion
    A valid metric name for this endpoint should include `headline` as the metric group part of the name.
```

File: metrics/api/views/audit/audit_api_timeseries.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+    - stratum `default`
+
+    This would be a **valid** request because the `metric` of `COVID-19_cases_casesByDay` relates to `Timeseries` data.
+
+    Note the key difference here being the 2nd part of the metric name:
+
+    *<topic>* *<metric group>* *<description>*
+
+    A valid metric name for this endpoint should **NOT** include `Headline` as the metric group part of the name.
```
Comment (A-Ashiq) [inline metrics/api/views/audit/audit_api_timeseries.py:67]:
```suggestion
    A valid metric name for this endpoint should **NOT** include `headline` as the metric group part of the name.
```

File: metrics/api/views/audit/serializers/audit_core_timeseries.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    stratum = serializers.SerializerMethodField()
+    refresh_date = serializers.SerializerMethodField()
+    embargo = serializers.SerializerMethodField()
+
+    class Meta:
+        model = CoreTimeSeries
+        fields = "__all__"
+
+    @classmethod
+    def get_topic(cls, obj) -> str:
```
Comment (A-Ashiq) [inline metrics/api/views/audit/serializers/audit_core_timeseries.py:33]: Same comment as before

File: tests/integration/metrics/api/views/audit/test_audit_core_headline.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        age = "all"
+
+        # When
+        path = (
+            f"{self.path}/{metric}/{geography_type}/{geography}/{age}/{sex}/{stratum}"
+        )
+        response: Response = client.get(path)
+
+        # Then
+        assert response.status_code == HTTPStatus.OK
```
Comment (A-Ashiq) [inline tests/integration/metrics/api/views/audit/test_audit_core_headline.py:43]: Same comment as before RE: checking for actual data

File: tests/integration/metrics/api/views/audit/test_audit_core_timeseries.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        age = "all"
+
+        # When
+        path = (
+            f"{self.path}/{metric}/{geography_type}/{geography}/{age}/{sex}/{stratum}"
+        )
+        response: Response = client.get(path)
+
+        # Then
+        assert response.status_code == HTTPStatus.OK
```
Comment (A-Ashiq) [inline tests/integration/metrics/api/views/audit/test_audit_core_timeseries.py:43]: As above

File: tests/unit/metrics/api/views/audit/serializers/test_audit_api_timeseries.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+from tests.fakes.managers.time_series_manager import FakeCoreTimeSeriesManager
+from tests.fakes.managers.topic_manager import FakeTopicManager
+from tests.fakes.models.metrics.topic import FakeTopic
+
+from metrics.data.models.api_models import APITimeSeries
+from metrics.api.views.audit.serializers import AuditAPITimeSeriesSerializer
+
+
+@pytest.fixture
+def mock_api_time_series() -> mock.MagicMock:
```
Comment (A-Ashiq) [inline tests/unit/metrics/api/views/audit/serializers/test_audit_api_timeseries.py:19]: Why not just use the fake api time series factory to get a fake version of the model?

File: tests/integration/metrics/api/views/audit/test_audit_api_timeseries.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+        age = "all"
+
+        # When
+        path = (
+            f"{self.path}/{metric}/{geography_type}/{geography}/{age}/{sex}/{stratum}"
+        )
+        response: Response = client.get(path)
+
+        # Then
+        assert response.status_code == HTTPStatus.OK
```
[the remark this answers is not in the document]
Comment (phill-stanley) [reply tests/integration/metrics/api/views/audit/test_audit_api_timeseries.py:43]: I've updated these endpoint tests now to include tests for returning the correct data and to test it returns data under embargo.

File: tests/unit/metrics/api/views/audit/serializers/test_audit_api_timeseries.py (hunk 10 above)
Comment (phill-stanley) [reply tests/unit/metrics/api/views/audit/serializers/test_audit_api_timeseries.py:19]: I've switched these over to use `FakeFactory...` for api, core timeseries and headline data in the unit tests for audit endpoints

File: tests/integration/metrics/api/views/audit/test_audit_api_timeseries.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        path = (
+            f"{self.path}/{metric}/{geography_type}/{geography}/{stratum}/{sex}/{age}"
+        )
+        response: Response = client.get(path)
+
+        # Then
+        assert response.status_code == HTTPStatus.OK
+        assert response.headers["Content-Type"] == "application/json"
+
+        assert len(response.data["results"]) == 1
```
Comment (A-Ashiq) [inline tests/integration/metrics/api/views/audit/test_audit_api_timeseries.py:61]: Why is this only returning the 1 record? Surely we want all the records for that data slice?

File: tests/integration/metrics/api/views/audit/test_audit_core_headline.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+        path = (
+            f"{self.path}/{metric}/{geography_type}/{geography}/{stratum}/{sex}/{age}"
+        )
+        response: Response = client.get(path)
+
+        # Then
+        assert response.status_code == HTTPStatus.OK
+        assert response.headers["Content-Type"] == "application/json"
+
+        assert len(response.data["results"]) == 1
```
Comment (A-Ashiq) [inline tests/integration/metrics/api/views/audit/test_audit_core_headline.py:58]: Same here, shouldn't this be all 3 records returned in the response?

**Review by A-Ashiq — APPROVED** [review]

**Pull request merged**